rust/src/delta.rs

```diff
@@ -613,6 +673,127 @@ impl std::fmt::Debug for DeltaTable {
     }
 }
 
+#[derive(thiserror::Error, Debug)]
+pub enum DeltaTransactionError {
+    #[error("Version already exists")]
+    VersionExists,
+
+    #[error("Maximum number of retries exceeded for committing the transaction")]
+    CommitRetriesExceeded,
+
+    #[error("RecordBatch is missing partition column in Delta schema.")]
+    MissingPartitionColumn,
+
+    #[error("Commit failed: {source}")]
+    Storage { source: StorageError },
+
+    #[error("DeltaTable interaction failed: {source}")]
+    DeltaTable { source: DeltaTableError },
+}
+
+impl From<StorageError> for DeltaTransactionError {
+    fn from(error: StorageError) -> Self {
+        match error {
+            StorageError::AlreadyExists => DeltaTransactionError::VersionExists,
+            _ => DeltaTransactionError::Storage { source: error },
+        }
+    }
+}
+
+impl From<DeltaTableError> for DeltaTransactionError {
+    fn from(error: DeltaTableError) -> Self {
+        DeltaTransactionError::DeltaTable { source: error }
+    }
+}
+
+pub struct DeltaTransaction<'a> {
+    delta_table: &'a mut DeltaTable,
+}
+
+impl<'a> DeltaTransaction<'a> {
+    pub fn new(delta_table: &'a mut DeltaTable) -> Self {
+        DeltaTransaction { delta_table,}
+    }
+
+    pub async fn commit_all(&mut self, actions: &[Action], _operation: Option<DeltaOperation>) -> Result<DeltaDataTypeVersion, DeltaTransactionError> {
+        // TODO: stubbing `operation` parameter (which will be necessary for writing the CommitInfo action), but leaving it unused for now.
+        // `CommitInfo` is a fairly dynamic data structure so we should work out the data structure approach separately.
+
+        // TODO: calculate isolation level to use when checking for conflicts.
+        // Leaving conflict checking unimplemented for now to get the "single writer" implementation off the ground.
+        // Leaving some commmented code in place as a guidepost for the future.
+
+        // let no_data_changed = actions.iter().all(|a| match a {
+        //     Action::add(x) => !x.dataChange,
+        //     Action::remove(x) => !x.dataChange,
+        //     _ => false,
+        // });
+        // let isolation_level = if no_data_changed {
+        //     IsolationLevel::SnapshotIsolation
+        // } else { 
+        //     IsolationLevel::Serializable
+        // };
+
+        // TODO: create a CommitInfo action and prepend it to actions.
+
+        // try to commit in a loop in case other writers write the next version first
+        let version = self.try_commit_loop(actions).await?;
+
```

Comment:
perhaps worth adding a note for future optimization, after the commit, we already have the new version and transaction log in memory, so we don't need to call update to fetch it from remote store.

Reply:
good call. added note.